Add cutlists to blocks/slopers

BenJamesBen opened the pull request:
PR to add cutlist info to the block/sloper designs. It is important to do this first, before adding cutlists to other designs that inherit from them.

- I chose to add a minimal amount of `addCut()`s, only adding code where needed to produce the correct cutting layout. (The alternative I considered was adding explicit cutlist code to every part, even when not required, in the name of potential clarity.)
- Some parts have a `removeCut()` that was needed to clear an inherited cutlist. I elected to not add a comment explaining these situations.
- I also considered adding an explicit `removeCut()` to every inheriting part regardless of whether it was needed, but I chose not to do so.
- I added a missing `cutonfoldAndGrainline` to Breanna's front. The foldline and grainline are mentioned in the instructions and code, so it appears that the omission was accidental. 
- I added a matching `grainline` to Breanna's back, assuming that it was also intended (given that the front is aligned on grainline) but accidentally omitted.

![Screenshot 2023-04-26 at 7 08 35 PM](https://user-images.githubusercontent.com/109869956/234743035-2fdd6607-3cd9-462f-95b7-9a17b74a4cf4.png)
![Screenshot 2023-04-26 at 7 09 09 PM](https://user-images.githubusercontent.com/109869956/234743045-b0e3b1a1-7b0d-417f-a93f-5aef4d9dae90.png)
![Screenshot 2023-04-26 at 7 08 15 PM](https://user-images.githubusercontent.com/109869956/234743078-526bf98e-eef5-4120-8c8c-5824fc74be3b.png)
![Screenshot 2023-04-26 at 7 09 26 PM](https://user-images.githubusercontent.com/109869956/234743098-e83d0025-ecd7-4083-b55a-432ac55310b1.png)
![Screenshot 2023-04-26 at 7 10 07 PM](https://user-images.githubusercontent.com/109869956/234743109-e1b72bf4-a501-468b-994b-47a78d0792c2.png)
![Screenshot 2023-04-26 at 7 17 31 PM](https://user-images.githubusercontent.com/109869956/234743121-1edf5b6e-1497-4c54-a81a-b13b3c885eee.png)

joostdecock commented:
Do we have enough consensus on this to go ahead with this?
I seem to remember there was a discussion on this during the last call, and I'm not sure it was entirely resolved. (bad memory)

eriese commented:
Up to @BenJamesBen. If we end up deciding cutlists should not be inherited, nothing in this PR will break just because it becomes unnecessary.

It's just whether you want to move things out of `if (complete)` in the same PR or not

BenJamesBen commented:
I moved `cutonfold` and `grainline` macros out of the `if (complete)` blocks. The PR is ready to merge.

(Please note that moving the macros isn't always a simple operation. The change actually broke Noble when `complete = false` because the macros in Noble's back parts depended upon `points.grainlineFrom` and `points.grainlineTo` being inherited from Bella's back from a previous `grainline` macro. Previously it was guaranteed that the points existed, back when all the macros were in `if (complete)`. With the change, the points don't exist when `complete = false`. To fix Noble, I added code to create the points.)

(A best practice is to test patterns to verify they still work with both `complete = true` and `complete = false` after moving macros out of `if (complete)`.)

(When re-creating the grainline points for Noble, I took the opportunity to move the backoutside part grainline to a better location for a greater range of measurements.)

Before:
![Screenshot 2023-04-27 at 2 55 10 PM](https://user-images.githubusercontent.com/109869956/234999653-a0b1244b-1bcd-468a-aab6-5ec5628aa12c.png)

After:
![Screenshot 2023-04-27 at 2 55 23 PM](https://user-images.githubusercontent.com/109869956/234999666-f8a18e8c-5712-4877-ac63-e12f304aa6ec.png)

eriese commented:
> I moved `cutonfold` and `grainline` macros out of the `if (complete)` blocks. The PR is ready to merge.
> 
> (Please note that moving the macros isn't always a simple operation. The change actually broke Noble when `complete = false` because the macros in Noble's back parts depended upon `points.grainlineFrom` and `points.grainlineTo` being inherited from Bella's back from a previous `grainline` macro. Previously it was guaranteed that the points existed, back when all the macros were in `if (complete)`. With the change, the points don't exist when `complete = false`. To fix Noble, I added code to create the points.)

Sounds like maybe we should be making the points and perhaps even the paths, but then hiding the paths if `complete === false`

> (A best practice is to test patterns to verify they still work with both `complete = true` and `complete = false` after moving macros out of `if (complete)`.)

good idea. We should add it to the standard tests

BenJamesBen commented:
> Sounds like maybe we should be making the points and perhaps even the paths, but then hiding the paths if complete === false

The documentation for `cutonfold` and `grainline` currently says that the paths and points are created when the macros are used. It does not mention `complete` as a condition.

We could either change the behavior of the macros as suggested above, or we could change the documentation to describe the current behavior. Since grainline and foldline information is something that we want to be inherited regardless of `complete` and regardless of whether the markings are drawn on the pattern, it sounds as if the first option should be done?

joostdecock commented:
I think it makes sense to let the macros take `complete` into account but developers should be able to force the to do it regardless
